sdk/cosmos/azure_data_cosmos/src/location_cache.rs

```diff
@@ -0,0 +1,570 @@
+        self.locations_info.write_endpoints = self.get_preferred_available_endpoints(
+            &self.locations_info.available_write_endpoints_by_location,
+            RequestOperation::Write,
+            &self.default_endpoint,
+        );
+
+        self.locations_info.read_endpoints = self.get_preferred_available_endpoints(
+            &self.locations_info.available_read_endpoints_by_location,
+            RequestOperation::Read,
+            &self.default_endpoint,
+        );
+
+        self.refresh_stale_endpoints();
+    }
+
+    fn get_endpoints_by_location(
+        &mut self,
+        locations: HashMap<String, String>,
+    ) -> (HashMap<String, String>, Vec<String>) {
+        // Separates locations into a hashmap and list
+        let mut endpoints_by_location: HashMap<String, String> = HashMap::new();
+        let mut parsed_locations: Vec<String> = Vec::new();
+
+        for (location, endpoint) in locations {
+            if !location.is_empty() {
+                endpoints_by_location.insert(location.clone(), endpoint.clone());
+                parsed_locations.push(location);
+            }
+        }
+
+        (endpoints_by_location, parsed_locations)
+    }
+
+    fn get_preferred_available_endpoints(
+        &self,
+        endpoints_by_location: &HashMap<String, String>,
+        request: RequestOperation,
+        default_endpoint: &str,
+    ) -> Vec<String> {
+        let mut endpoints: Vec<String> = Vec::new();
+        let mut unavailable_endpoints: Vec<String> = Vec::new();
+
+        for location in &self.locations_info.preferred_locations {
+            // Checks if preferred location exists in endpoints_by_location
+            if let Some(endpoint) = endpoints_by_location.get(location) {
+                // Check if endpoint is available, if not add to unavailable_endpoints
+                // If it is then add to endpoints
+                if !self.is_endpoint_unavailable(endpoint, request) {
+                    endpoints.push(endpoint.clone());
+                } else {
+                    unavailable_endpoints.push(endpoint.clone());
+                }
+            }
+        }
+
+        // Add unavailable endpoints to end of endpoints lists
+        for endpoint in unavailable_endpoints {
+            endpoints.push(endpoint);
+        }
+
+        // If no preferred locations were found, use the default endpoint
+        if endpoints.is_empty() {
+            endpoints.push(default_endpoint.to_string());
+        }
+
+        endpoints
+    }
+
+    fn refresh_stale_endpoints(&mut self) {
+        let mut location_unavailability_info_map =
+            self.location_unavailability_info_map.lock().unwrap();
+
+        // Removes endpoints that have not been checked in the last 5 minutes
+        location_unavailability_info_map.retain(|_, info| {
+            info.last_check_time.elapsed().unwrap_or_default()
+                <= std::time::Duration::from_secs(300)
```

Comment:
Put the "freshness" duration in a constant at the top of the file so we know what it is and what it's for. Also (and this is just personal preference), I'll often write this as a math expression rather than just hardcoding the number, since it makes it clearer how it's derived. The compiler will take care of doing the math at compile time so there's no runtime impact:

```
const ENDPOINT_FRESHNESS_DURATION: std::time::Duration = std::time::Duration::from_secs(5 * 60);
```

When you see `from_secs(5 * 60)` it's easier to know that's 5 minutes (and then, honestly, you don't need the comment)